virtualizarr/storage/_obstore.py

```diff
@@ -0,0 +1,246 @@
+from __future__ import annotations
+
+import pickle
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any
+
+from xarray import DataArray, Dataset
+from zarr.abc.store import (
+    ByteRequest,
+    OffsetByteRequest,
+    RangeByteRequest,
+    Store,
+    SuffixByteRequest,
+)
+from zarr.core.buffer import Buffer
+from zarr.core.buffer.core import BufferPrototype
+
+from virtualizarr.storage.common import (
+    find_matching_store,
+    get_zarr_metadata,
+    list_dir_from_xr_obj,
+    parse_manifest_index,
+)
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator, Iterable
+    from typing import Any
+
+    from obstore.store import ObjectStore as _UpstreamObjectStore
+    from zarr.core.buffer import BufferPrototype
+    from zarr.core.common import BytesLike
+
+
+__all__ = ["VirtualObjectStore"]
+
+_ALLOWED_EXCEPTIONS: tuple[type[Exception], ...] = (
+    FileNotFoundError,
+    IsADirectoryError,
+    NotADirectoryError,
+)
+
+
+class VirtualObjectStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
+
+    Parameters
+    ----------
+    stores : dict[prefix, obstore.store.ObjectStore]
+        A mapping of url prefixes to obstore Store instances set up with the proper credentials.
+
+    Warnings
+    --------
+    VirtualObjectStore is experimental and subject to API changes without notice. Please
+    raise an issue with any comments/concerns about the store.
+
+    Notes
+    -----
+    Modified from https://github.com/zarr-developers/zarr-python/pull/1661
+    """
+
+    def __eq__(self, value: object):
+        NotImplementedError
+
+    def __init__(
+        self,
+        xr_obj: DataArray | Dataset,
+        stores: dict[str, _UpstreamObjectStore],
+    ) -> None:
+        import obstore as obs
+
+        # TODO: Support DataArray, Dataset, or DataTree across all methods
+        for store in stores.values():
+            if not isinstance(
+                store,
+                (
+                    obs.store.AzureStore,
+                    obs.store.GCSStore,
+                    obs.store.HTTPStore,
+                    obs.store.S3Store,
+                    obs.store.LocalStore,
+                    obs.store.MemoryStore,
+                ),
+            ):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        super().__init__(read_only=True)
+        self.stores = stores
+        self.xr_obj: DataArray | Dataset = xr_obj
```

Comment:
You can add type hints for these in the class definition, outside of any of the methods.